### PR TITLE
AK: Add `count{,_leading,_trailing}_{ones,zeros}`

### PR DESCRIPTION
**Pull request by mhermier**
In an effort to reduce the usage of compiler intrinsic in the code, move their usage to a single place.

### CODE REVIEW

**Comment by Dexesttp** [comment]
Right now, this PR seems to add dead (unused) code to the project, so it's not too interesting without any other code actually using it somewhere :)

Maybe if there's somewhere where these should be used, you could add that change to this PR as well?

**Comment by mhermier** [comment]
I can do that, thought I would prefer some people who can more properly test their changes to do the commits. As these functions can produce different results based on the type they are feed with, I don't want to introduce bugs I can't properly test.

**Comment by Hendiadyoin1** [comment]
Both functions can already be found I the project, for example in AK/Math.h

I think they are in other places too

File: AK/CountOnes.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        return __builtin_popcountll(val);
+#endif
+
+    size_t count = 0;
+    do {
+        if (val & 1) {
+            count++;
+        }
+    } while (val >>= 1);
+    return count;
```
Comment (Hendiadyoin1) [inline AK/CountOnes.h:34]: there most definitely is a better way to do this...

[here for example](https://en.wikichip.org/wiki/population_count)

File: AK/CountLeadingZeros.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        return __builtin_clzll(val);
+#endif
+
+    for (size_t count = 0; count < T_bit; ++count) {
+        if ((val >> (T_bit - count - 1)) & 1) {
+            return count;
+        }
+    }
+    VERIFY_NOT_REACHED();
+    return T_bit;
```
[the remark this answers is not in the document]
Comment (mhermier) [reply AK/CountLeadingZeros.h:40]: Sure but I prefer to be safe than sorry, and the compiler will probably refuse to compile without the trailing return.

File: AK/CountLeadingZeros.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+template<Integral T>
+ALWAYS_INLINE constexpr size_t count_leading_zeros(T val)
+{
+    constexpr size_t char_bit = 8;
+    constexpr size_t T_bit = sizeof(T) * char_bit;
+
+    if (val == 0)
+        return T_bit;
+
+#if defined(__GNUC__) || defined(__clang__)
```
[the remark this answers is not in the document]
Comment (mhermier) [reply AK/CountLeadingZeros.h:25]: True but this code was copy/pasted from `count_trailing_zeroes_32(_safe)` before I fixed it, so it has the same DNA.

File: AK/CountOnes.h (hunk 1 above)
Comment (mhermier) [reply AK/CountOnes.h:34]: Probably, but it is functional, and I'm chasing an existing bug for now...

File: AK/CountLeadingZeros.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhermier) [reply AK/CountLeadingZeros.h:40]: Seems I found another issue in the headers, I based my assumption that `VERIFY_NOT_REACHED()` would work with a conformant `<assert.h>`. In that case `assert(false)` does not return only when `NDEBUG` is not defined. But `Userland/Libraries/LibC/assert.h` does `#    define VERIFY_NOT_REACHED() _abort()` so it should produce a compile time error where `VERIFY_NOT_REACHED` is redefined...

**Comment by mhermier** [comment]
Will need to revamp my changes to adapt to "BuiltinWrappers.h".

**Pull request closed without merging**